### PR TITLE
Add support for bz2 compression in python

### PR DESCRIPTION
**Pull request by tim1357**
This is my first contribution to any open source project ever, so I have no idea if I'm doing this right.

### CODE REVIEW

File: sdks/python/apache_beam/io/fileio_test.py [hunk 1]
```diff
@@ -199,7 +250,7 @@ def test_read_entire_file_zlib_large(self):
         read_lines.append(line)
     self.assertEqual(read_lines, lines)
 
-  def test_skip_entire_file_gzip(self):
```
Comment (tim1357) [inline sdks/python/apache_beam/io/fileio_test.py]: test_skip_entire_file_gzip should be test_skip_entire_file_zlib

**Comment by tim1357** [comment]
If this goes through I'll start working on splittable bz2

**Comment by tim1357** [comment]
I'm not sure why Travis has stalled. For the record travis said it was fine on [my forked branch](https://travis-ci.org/tim1357/incubator-beam/builds/162886767). Is there a way to get it to try to build it again?

**Comment by tim1357** [comment]
@robertwb

**Comment by tim1357** [comment]
R: @jbonofre

File: sdks/python/apache_beam/io/fileio.py [hunk 2]
```diff
@@ -595,14 +602,21 @@ def __init__(self,
     self._compression_type = compression_type
 
     if self._readable():
-      self._decompressor = zlib.decompressobj(self._type_mask[compression_type])
+      if self._compression_type == CompressionTypes.BZIP2:
+        self._decompressor = bz2.BZ2Decompressor()
+      else:
+        self._decompressor = zlib.decompressobj(\
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio.py]: No need for a trailing backspace.

File: sdks/python/apache_beam/io/fileio.py [hunk 3]
```diff
@@ -91,14 +95,17 @@ def is_valid_compression_type(cls, compression_type):
   def mime_type(cls, compression_type, default='application/octet-stream'):
     mime_types_by_compression_type = {
         cls.GZIP: 'application/x-gzip',
-        cls.ZLIB: 'application/octet-stream'
+        cls.ZLIB: 'application/octet-stream',
+        cls.BZIP2: 'application/octet-stream'
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio.py]: (minor) I'd add a trailing comma here so if more formats are added, only one line changes.

File: sdks/python/apache_beam/io/fileio_test.py [hunk 4]
```diff
@@ -149,6 +150,56 @@ def test_read_entire_file_gzip_large(self):
         read_lines.append(line)
     self.assertEqual(read_lines, lines)
 
+  def test_read_entire_file_bzip(self):
+    lines = ['First', 'Second', 'Third']
+    compressor = bz2.BZ2Compressor(8)
+    data = compressor.compress('\n'.join(lines)) +compressor.flush()
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio_test.py]: space after the +

File: sdks/python/apache_beam/io/fileio.py [hunk 5]
```diff
@@ -27,6 +27,7 @@
 import threading
 import time
 import zlib
+import bz2
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio.py]: Can we assume bz2 is available on all target platforms? @charlesccychen

**Comment by robertwb** [comment]
Thanks for taking the time to do this! I see @jbonofre is officially signed up to review, but I had some minor comments.

**Comment by katsiapis** [comment]
There are a few enhancements needed and parts missing from the current PR, in order for BZIP2 support to be complete.

These are all actually taken care of in https://github.com/apache/incubator-beam/pull/1053
Does it make sense to proceed with that PR instead of the current one?

File: sdks/python/apache_beam/io/fileio.py (hunk 3 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: Le'ts use 'application/x-bz2' for the case of BZIP2.

File: sdks/python/apache_beam/io/fileio.py (hunk 5 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: Let's alpha-sort this. bz2 should appear in the top.

File: sdks/python/apache_beam/io/fileio.py [hunk 6]
```diff
@@ -595,14 +602,21 @@ def __init__(self,
...
+                            self._type_mask[compression_type])
     else:
       self._decompressor = None
 
     if self._writeable():
-      self._compressor = zlib.compressobj(zlib.Z_DEFAULT_COMPRESSION,
-                                          zlib.DEFLATED,
-                                          self._type_mask[compression_type])
+      if self._compression_type == CompressionTypes.BZIP2:
+        self._compressor = bz2.BZ2Compressor(9)
```
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: I recommend going with default compression level (ie don't specify one in the constructor).
Just bz2.BZ2Compressor()

File: sdks/python/apache_beam/io/fileio_test.py [hunk 7]
```diff
@@ -25,6 +25,7 @@
 import tempfile
 import unittest
 import zlib
+import bz2
```
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio_test.py]: See other comment regarding alpha-sort.

File: sdks/python/apache_beam/io/fileio_test.py [hunk 8]
```diff
@@ -149,6 +150,56 @@ def test_read_entire_file_gzip_large(self):
         read_lines.append(line)
     self.assertEqual(read_lines, lines)
 
+  def test_read_entire_file_bzip(self):
+    lines = ['First', 'Second', 'Third']
+    compressor = bz2.BZ2Compressor(8)
```
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio_test.py]: Perhaps remove the 8 compression level? Similarly for all occurrences below.

File: sdks/python/apache_beam/io/fileio_test.py [hunk 9]
```diff
@@ -149,6 +150,56 @@ def test_read_entire_file_gzip_large(self):
         read_lines.append(line)
     self.assertEqual(read_lines, lines)
 
+  def test_read_entire_file_bzip(self):
```
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio_test.py]: Could we do a rename from bzip -> bzip2 for all those tests?

File: sdks/python/apache_beam/io/fileio.py (hunk 5 above)
Comment (tim1357) [inline sdks/python/apache_beam/io/fileio.py]: [done](https://github.com/apache/incubator-beam/pull/1007/commits/41e2c4a7b196b3ac26a6c3c80f109e0ea996a4d6#diff-b5a9043bc2c548bb7e1b0fca276f5329R21)

File: sdks/python/apache_beam/io/fileio.py (hunk 3 above)
Comment (tim1357) [inline sdks/python/apache_beam/io/fileio.py]: [done](https://github.com/apache/incubator-beam/pull/1007/commits/41e2c4a7b196b3ac26a6c3c80f109e0ea996a4d6#diff-b5a9043bc2c548bb7e1b0fca276f5329R99)

File: sdks/python/apache_beam/io/fileio.py (hunk 2 above)
Comment (tim1357) [inline sdks/python/apache_beam/io/fileio.py]: [done](https://github.com/apache/incubator-beam/pull/1007/commits/41e2c4a7b196b3ac26a6c3c80f109e0ea996a4d6#diff-b5a9043bc2c548bb7e1b0fca276f5329R608)

File: sdks/python/apache_beam/io/fileio_test.py (hunk 4 above)
Comment (tim1357) [inline sdks/python/apache_beam/io/fileio_test.py]: [done](https://github.com/apache/incubator-beam/pull/1007/commits/41e2c4a7b196b3ac26a6c3c80f109e0ea996a4d6#diff-a28e0556968b45ca2f326d54a3ff75aaR156)

File: sdks/python/apache_beam/io/fileio_test.py (hunk 9 above)
Comment (tim1357) [inline sdks/python/apache_beam/io/fileio_test.py]: done

File: sdks/python/apache_beam/io/fileio.py (hunk 6 above)
Comment (tim1357) [inline sdks/python/apache_beam/io/fileio.py]: done

File: sdks/python/apache_beam/io/fileio_test.py (hunk 8 above)
Comment (tim1357) [inline sdks/python/apache_beam/io/fileio_test.py]: done

**Comment by tim1357** [comment]
@robertwb Ive implemented the changes suggested in the pr. Is there anything else you'd like to see?

**Comment by robertwb** [comment]
The travis error seems to be due to the ongoing issues we're having with OS X tox. I've merged this, you can close this PR now. 

Thanks again for your contribution!

**Comment by katsiapis** [comment]
Thanks @tim1357 for the contribution and Robert for the review. LGTM as well.

@tim1357, to avoid any confusion/conflicts, could you do any extra development on top of https://github.com/apache/incubator-beam/pull/1053 (which I will work on resolving/merging soon)?

**Comment by robertwb** [comment]
@tim1357 , could you manually close this request now that it's been merged?

**Pull request closed without merging**